### PR TITLE
SQL: (0.5) Versioned partitions

### PR DESCRIPTION
**Pull request by midenok**
* one `AS OF NOW`, multiple `VERSIONING` partitions;
* rotation of `VERSIONING` partitions by record count, time period;
* rotation is multi-threaded;
* conventional subpartitions as bottom level for versioned partitions;
* `DEFAULT` keyword selects first `VERSIONING` partition.

### CODE REVIEW

**Comment by midenok** [comment]
Please, concentrate on review now. Finishing it on 13th evening would be good.

File: sql/partition_info.cc [hunk 1]
```diff
@@ -1058,6 +1072,215 @@ bool partition_info::has_unique_name(partition_element *element)
...
+      if (!vers_info->hist_part && el->id == vers_info->hist_default)
+      {
+        vers_info->hist_part= el;
+      }
+      if (is_create_table_ind || (
+        table->s->free_parts_init &&
+        !vers_limit_exceed(el) &&
+        !vers_interval_exceed(el)))
+      {
+        table->s->free_parts.push_back((void *) el->id, &table->mem_root);
```
Comment (midenok) [inline sql/partition_info.cc]: Bug: `table->s->free_parts` lives longer than `table->mem_root`

File: sql/sql_partition.cc [hunk 2]
```diff
@@ -8445,4 +8580,5 @@ uint get_partition_field_store_length(Field *field)
     store_length+= HA_KEY_BLOB_LENGTH;
   return store_length;
 }
+
```
Comment (midenok) [inline sql/sql_partition.cc]: Junk.

File: sql/handler.h [hunk 3]
```diff
@@ -4223,6 +4222,13 @@ class handler :public Sql_alloc
   virtual void set_lock_type(enum thr_lock_type lock);
 
   friend enum icp_result handler_index_cond_check(void* h_arg);
+
+  virtual bool versioned() const
+  { DBUG_ASSERT(ht); return partition_ht()->flags & HTON_SUPPORTS_SYS_VERSIONING; }
+  virtual ha_rows part_recs_slow(void *hist_part)
```
Comment (kevgs) [inline sql/handler.h]: Rename `hist_part` to `_part_elem` as [here](https://github.com/natsys/mariadb_10.2/pull/93/files#diff-3de26a45d78578fa2c27154e2c264a39R1290)

File: sql/ha_partition.h [hunk 4]
```diff
@@ -1282,6 +1282,38 @@ class ha_partition :public handler
...
+    DBUG_ASSERT(m_part_info);
+    uint32 sub_factor= m_part_info->num_subparts ? m_part_info->num_subparts : 1;
+    uint32 part_id= part_elem->id * sub_factor;
+    uint32 part_id_end= part_id + sub_factor;
+    DBUG_ASSERT(part_id_end <= m_tot_parts);
+    ha_rows part_recs= 0;
+    for (; part_id < part_id_end; ++part_id)
+    {
+      handler *file= m_file[part_id];
+      DBUG_ASSERT(bitmap_is_set(&(m_part_info->read_partitions), part_id));
```
Comment (kevgs) [inline sql/ha_partition.h]: Unneeded parentheses around `m_part_info->read_partitions`.

File: sql/handler.cc [hunk 5]
```diff
@@ -6491,7 +6491,7 @@ static bool create_string(MEM_ROOT *mem_root, String **s, const char *value)
   return *s == NULL;
 }
 
-static bool create_sys_trx_field(THD *thd, const char *field_name,
+static bool vers_create_sys_field(THD *thd, const char *field_name,
                                  Alter_info *alter_info, String **s,
```
Comment (kevgs) [inline sql/handler.cc]: One more space to indent after function renaming?

**Pull request merged**